### PR TITLE
Write about our continuous release process

### PR DESCRIPTION
**Pull request by alexwlchan**
I promised @DRMacIver a post about this at PyCon UK… in 2017. That’s nearly 60 releases ago! And then I was CamPUG this evening, and Tibbs was asking for the post about this, so here we go.

Current status: first draft written at the end of a long day. Definitely needs a good edit, and I’d be interested in hearing whether there are benefits I’ve forgotten that I should write about.

### CODE REVIEW

**Comment by alexwlchan** [comment]
Okay, I think this is ready for somebody other than me to review.

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: Due to an annoying limitation of Jekyll on Github Pages, our tags are not entirely freeform and new tags need to be added in https://github.com/HypothesisWorks/HypothesisWorks.github.io/tree/master/articles

(there should probably be a script for this)

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: Huh. I didn't know about this markdown link syntax.

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: Minor historical note: In earlier history, Hypothesis releases were actually *really* fast - possibly faster than they are today due to much more limited testing. The problem seemed to be that as soon as the release cycle stopped being virtually instantaneous it ground to a halt. I'm not sure exactly how and why this happened, but I think it was partly due to improved testing and process, so the previous YOLO deployments stopped being viable.

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: Maybe clarify that the PhD hadn't actually started then? e.g. "groundwork for David's upcoming (now started) PhD"? (Not at all wed to that wording)

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: Possibly worth noting that this is the second gen automation system - the first required updating the changelog yourself and was awful.

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: It might be worth noting that the code is a little hairy and will probably require some adaptation...

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: Another unmentioned advantage is that I *really* hated doing these manual releases! Collecting the changelog manually was really annoying, and we didn't really have a previously good system of change log automation.

File: _posts/2018-02-12-continuous-releases.md
Comment (DRMacIver) [inline _posts/2018-02-12-continuous-releases.md]: Maybe worth mentioning that you can create a `RELEASE.rst` without source changes and we sometimes do for e.g. making `setup.py` changes (which maybe we should just add to the checks TBH) and test changes for the benefit of downstream packagers?

**Review by DRMacIver — CHANGES_REQUESTED** [review]
Generally looks 💯. Have added some comments on things I think could use clarifying or elaborating on.

Would it be worth talking about some of the limitations and edge cases?

e.g.

* This blocks merges of new features until the old one has been released, by virtue of the conflicting adds.
* If a non-feature PR has been merged during the build, the push of the new commit will fail and the deployment will fail, but the build of the new merged version of master will deploy it.

File: _posts/2018-02-12-continuous-releases.md
Comment (alexwlchan) [reply _posts/2018-02-12-continuous-releases.md]: Yeah, it's quite nice for including links without cluttering up your prose with URLs.

**Comment by alexwlchan** [comment]
@DRMacIver I think I’ve addressed all your comments, but I want to go over it with a red pen and paper before the next review. I'll let you know when it's ready for a second review.

**Comment by dchudz** [comment]
I wonder if it might be worth saying something about how this workflow interacts with supporting old versions. (Hypothesis doesn't ever add a release to old major or minor versions, right?)

I think a project wouldn't have much trouble amending this workflow to support e.g. one old major version. But doing much more than that seems like it would be bad news. This all seems to work really well for Hypothesis but I do wonder about projects with e.g. critical security patches. 

It's a great post even without saying anything about this - I'm not really saying you *should* address it. I guess I'm just mentioning the question because it's one I think readers will ask.

**Comment by DRMacIver** [comment]
> I wonder if it might be worth saying something about how this workflow interacts with supporting old versions. (Hypothesis doesn't ever add a release to old major or minor versions, right?)

This isn't a thing we do, so it's not ever really came up.

If anything though I imagine it would make it easier (modulo some wrangling of git): If you apply a patch with a `RELEASE.rst` to an older version, the version number automation would automatically do the right thing.

Hard to say for sure though. I'd be reluctant to suggest much about this without having actually tried such a workflow.

**Comment by dchudz** [comment]
> I'd be reluctant to suggest much about this without having actually tried such a workflow.

Yeah, that makes sense. Sorry, I think I really should have just suggested saying something like that: "We like this workflow, but haven't tried it for ... so your mileage may vary." But I guess it kinda goes without saying too. Maybe it wasn't worth mentioning, sorry!

**Comment by alexwlchan** [comment]
> I wonder if it might be worth saying something about how this workflow interacts with supporting old versions.

I think the answer is “it doesn’t, at least not in its current form”. 😜 

I’ll put a sentence about it in the “this won’t necessarily work for larger projects” paragraph – I think this is a good point, and we don't need to offer solutions to say that it's a problem.

**Comment by alexwlchan** [comment]
Okay, markups made. I think this is ready for another look @DRMacIver.

**Review by DRMacIver — APPROVED** [review]

**Comment by DRMacIver** [comment]
Looks great!

I'll leave it up to you to merge whenever you want. Sorry for the delayed rereview.

**Comment by alexwlchan** [comment]
I’ll need to fix the date (we should totes do continuous deployment here as well!) when I get home; I’ll probably post on Monday morning.

**Comment by alexwlchan** [comment]
@DRMacIver Also it looks like I don't have write access on this repo.

**Comment by alexwlchan** [comment]
@DRMacIver I’ve bumped the post date. At this point either you can give me write access, or merge it immediately.

**Comment by DRMacIver** [comment]
> @DRMacIver I’ve bumped the post date. At this point either you can give me write access, or merge it immediately.

I went with "both"

**Pull request merged**
